Handle packages with duplicate names (resolve #21)

khanhduong95 opened the pull request:
Resolve issue #21 

Automatically detect conflicts between different packages sharing the same name (for example: "app/http/clients" vs "app/grpc/clients") and generate unique aliases

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/elliotchance/dingo/22)
<!-- Reviewable:end -->

Commit:
Sort arguments

khanhduong95 committed:
Merge branch 'elliotchance:master' into master

claude committed:
Add auto-resolution for duplicate package names

This implementation adds conflict detection and resolution when multiple
dependencies share the same package name but have different parent paths
(e.g., "app/http/clients" vs "app/grpc/clients").

Key changes:
- Added imports.go with CollectAndResolveImports() to detect conflicts
- When conflicts are detected, unique names are generated by incorporating
  parent path segments (e.g., "http_clients" and "grpc_clients")
- Updated Type methods to support resolved import maps
- Added context-aware package prefix replacement in expressions
- All existing tests pass with backward compatibility maintained

Example:
  Services with types:
    - app/http/clients.Client
    - app/grpc/clients.Client

  Generated imports:
    import (
        grpc_clients "app/grpc/clients"
        http_clients "app/http/clients"
    )

claude committed:
Clean up unnecessary and redundant comments

elliotchance commented:
Please provide a better description of what the bug is and how it is fixed with this change.

khanhduong95 commented:
> Please provide a better description of what the bug is and how it is fixed with this change.

I've updated

elliotchance commented:
I still don't understand what this proposed change is actually fixing? If you have two interfaces in different packages, why can't you simply use the FQN as described in the README?

khanhduong95 commented:
> I still don't understand what this proposed change is actually fixing? If you have two interfaces in different packages, why can't you simply use the FQN as described in the README?

I'm referring to case where 2 different packages have the same name (but different parent folders), resulting in alias conflict in generated container code
For example: 2 interfaces
- foo1/bar.Interface1
- foo2/bar.Interface2
the resulting container code will have the imports like this
`import (`
`    bar "foo1/bar"`
`    bar "foo2/bar" // CONFLICT`
`)`

elliotchance commented:
I haven't looked through the full PR, but it looks to be too many changes for such a fix. I will accept that the `import` should allow for two words that use the same import alias form in go:

```yaml
import:
  - httpclients app/http/clients
  - grpcclients app/grpc/clients
  - some/other/package
```

claude committed:
Remove unused replacePackagePrefixes function

claude committed:
Refactor: eliminate *WithMap method duplication

Replace duplicate methods with single methods accepting optional importMap:
- LocalPackageName(importMap) instead of separate *WithMap version
- LocalEntityType(importMap) instead of separate *WithMap version
- ContainerFieldType(services, importMap) instead of separate *WithMap version
- And all related methods

This eliminates ~150 lines of duplication while maintaining all functionality.
All call sites updated to pass nil when importMap not needed.
